pkg/controller/seed-controller-manager/constraint-controller/controller_test.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2021 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package seedconstraintsynchronizer
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	v1 "k8c.io/kubermatic/v2/pkg/api/v1"
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/handler/test"
+	kubermaticlog "k8c.io/kubermatic/v2/pkg/log"
+	"k8c.io/kubermatic/v2/pkg/util/workerlabel"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	fakectrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	constraintName = "constraint"
+	kind           = "RequiredLabel"
+)
+
+func TestReconcile(t *testing.T) {
+
+	workerSelector, err := workerlabel.LabelSelector("")
+	if err != nil {
+		t.Fatalf("failed to build worker-name selector: %v", err)
+	}
+
+	testCases := []struct {
+		name                 string
+		namespacedName       types.NamespacedName
+		requestName          string
+		expectedConstraint   *kubermaticv1.Constraint
+		expectedGetErrStatus metav1.StatusReason
+		seedClient           ctrlruntimeclient.Client
+	}{
+		{
+			name: "scenario 1: sync constraint to user cluster ns",
```

Comment:
not sure if this tests works correctly, in this first case you are getting and comparing to the same constraint which is generated in the `seedClient` as you are doing the GET below with the NamespacedName from it, you are not checking if the namespace is the proper one.

Similarly for tests 2 and 3 the reconcile doesnt even happen properly as you are giving it an empty NamespaceName.

You need to generate a cluster as well, a namespace for it and check if the sync happens properly

Reply:
done